libnetwork/internal/addrset/addrset.go

```diff
@@ -0,0 +1,198 @@
+// Package addrset implements a set of IP addresses.
+package addrset
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/netip"
+	"strings"
+
+	"github.com/docker/docker/libnetwork/bitmap"
+	"github.com/docker/docker/libnetwork/internal/netiputil"
+	"github.com/docker/docker/libnetwork/ipbits"
+)
+
+var (
+	// ErrBadAddress is returned when a requested address is not in the pool
+	ErrBadAddress = errors.New("address not in pool")
+	// ErrNotAvailable is returned when no more addresses are available to set
+	ErrNotAvailable = errors.New("address not available")
+	// ErrAllocated is returned when the specific address requested is already allocated
+	ErrAllocated = errors.New("address already allocated")
```

Comment:
I guess I was wondering if we have some option to alias some of the types so that we don't have to introduce yet-another set of errors 😅, e.g.


```suggestion
	// ErrNotAvailable is returned when no more addresses are available to set
	ErrNotAvailable = ipamapi.ErrNoAvailableIPs
	// ErrAllocated is returned when the specific address requested is already allocated
	ErrAllocated = ipamapi.ErrIPAlreadyAllocated
```

or

```suggestion
	// ErrNotAvailable is returned when no more addresses are available to set
	ErrNotAvailable = bitmap.ErrBitAllocated
	// ErrAllocated is returned when the specific address requested is already allocated
	ErrAllocated = bitmap.ErrNoBitAvailable
```

(but the bitmap errors don't implement `errdefs` types)

Reply:
The bitmap error texts talk about "bits" being available, but this new package is dealing with a set of addresses. So, bitmap's error messages didn't really make sense - and maybe we'll want to represent the address set differently at some point.

It didn't seem right for `AddrSet` to return IPAM errors, even though it's only used by `defaultipam` at the moment. But if you think it should, I can do that? (Could make package `addrset` internal to `defaultipam` so that it can't be used by anything else?)

Reply:
Yeah, it was mostly wishful thinking on this one; libnetwork had a bit of a tendency to define errors for everything, even for those not actually checked for; in some cases that led to tighter coupling than needed.

So I was looking if it would be possible somehow to get closer to our errdefs definitions for matching specific conditions, but otherwise keep things generic.

This is all internal, so not a huge deal (doesn't expand public APIs / contracts.